### ISSUE TITLE
How to get all layers in a aseprite file?

### ISSUE THREAD

**Issue by Kietyo** [korlibs/korge#1193]
Here's my aseprite:

<img width="960" alt="image" src="https://user-images.githubusercontent.com/5054996/211471924-521998e1-3817-40ef-ae71-328a8609de39.png">

There doesn't seem to be a way to get all layers in the file. When reading I only get the base layer:

<img width="481" alt="image" src="https://user-images.githubusercontent.com/5054996/211472387-3efcbfe8-5e3e-4ebc-8c07-48670ff9235d.png">

```
            val asp = resourcesVfs["tower_sprites.aseprite"].readImageDataContainer(ASE.toProps())

            println(asp)
            println(asp.imageDatas)
            println(asp.imageDatasByName)

            println(asp.imageDatas.first())
            println(asp.imageDatas.first().frames.first())

            image(asp.imageDatas.first().frames.first().bitmap) {
                scaleWhileMaintainingAspect(ScalingOption.ByWidthAndHeight(300.0, 300.0))
            }

```

Output of println statements:

```
com.soywiz.korim.format.ImageDataContainer@5404e08b
[ImageData([ImageFrame(BitmapSlice(null:SizeInt(width=32, height=32)), time=100ms, targetX=0, targetY=0, main=false)])]
{null=ImageData([ImageFrame(BitmapSlice(null:SizeInt(width=32, height=32)), time=100ms, targetX=0, targetY=0, main=false)])}
ImageData([ImageFrame(BitmapSlice(null:SizeInt(width=32, height=32)), time=100ms, targetX=0, targetY=0, main=false)])
ImageFrame(BitmapSlice(null:SizeInt(width=32, height=32)), time=100ms, targetX=0, targetY=0, main=false)
```

**Comment by Kietyo** [comment]
File

**Comment by jobe-m** [comment]
I guess you need to create "image" objects for each layer of a frame. Take a look here how it is done in imageAnimationView setFrame function: https://github.com/korlibs/korge/blob/109d582eb8b6810403f406ad28872b480bb9fb8b/korge/src/commonMain/kotlin/com/soywiz/korge/view/animation/ImageAnimationView.kt#L90

Something like that:

    asp.imageDatas.first().frames.first().layerData.fastForEach { image(it.layer) }

Or create directly an ImageAnimationView object. It will contain and display all (visible) layers of the aseprite file.
There is also the possibility to load only specific layers of an aseprite file with the readImageContainer() function. Description how to do that is here: https://github.com/korlibs/korge/blob/109d582eb8b6810403f406ad28872b480bb9fb8b/korim/src/commonMain/kotlin/com/soywiz/korim/format/ASE.kt#L237

.oO( I should document that in docs.korge.org )

**Comment by Kietyo** [comment]
Thanks! I understand a bit better now how it works. I wonder if the API could be improved such that it returns a model that's more specific to ASE rather than use the more generic ImageDataContainer. I will try to implement this in my own project for now.

**Comment by Kietyo** [comment]
Actually I have another issue I was wondering if you could advise me on.

So in my aseprite file I have a gun turret that looks like this:

<img width="960" alt="image" src="https://user-images.githubusercontent.com/5054996/211721065-5627d84e-8d9b-4b1d-b6a1-fa9fb57a8dbd.png">

I noticed that when I get the bitmap for that layer, it appears to "crop" to the region that is colored, rather than return a 32x32 bitmap:

<img width="481" alt="image" src="https://user-images.githubusercontent.com/5054996/211721510-264134f7-43a9-4483-b545-de6187a57f82.png">

I was wondering if there's a way to return an "uncropped" version?

I tried out permutations of the "disableSlicing" and "useSlicePosition" properties but they don't seem to do anything.

**Comment by jobe-m** [comment]
> I wonder if the API could be improved such that it returns a model that's more specific to ASE rather than use the more generic ImageDataContainer.

I don't think we should change the model data towards ASE since it would make Korge dependent on it. Better is to stay independent so that also other image formats can be supported more easily. But we can of course add new features which might come from the ASE format.

> I was wondering if there's a way to return an "uncropped" version?

I don't remember currently if the ASE file format saves images already cropped or if Korge upon loading crops the image data. But I remember that I added a "virtual rect and offset" which corresponds to the original image size (-> rect) and the position of the cropped image inside the original size (-> offset).
With that information Korge is able to place the cropped image "inside" of the original image size which you normally use in Aseprite (e.g. 32x32). So you do not have to deal with the inherent offset which comes with the cropped image.
I guess for that to work you need to use the ImageAnimationView.

**Comment by Kietyo** [comment]
Thanks for the advice!

> I don't think we should change the model data towards ASE since it would make Korge dependent on it.

My suggestion is not to change the model data towards ASE but create a new data model that is returned specifically for ASE that is self-documenting and matches what the user would expect. For my own project I will probably create a new data model and convert from the ImageDataContainer to that new data model for better readability/understandability.

**Comment by Kietyo** [comment]
Okay I got it!

So ImageFrameLayer.targetX and ImageFrameLayer.targetY stores the offset of the "cropped" bitmap, which you can use to create the full 32x32 bitmap:

```
            println(turretLayer.width)
            println(turretLayer.height)
            println(turretLayer.targetX)
            println(turretLayer.targetY)
```

```output
8
20
12
6
```
